frontend_server/lib/server.dart

```diff
@@ -0,0 +1,207 @@
+  /// Assuming some Dart program was previously compiled, recompile it again
+  /// taking into account some changed(invalidated) sources.
+  Future<Null> recompileDelta();
+  /// Accept results of previous compilation so that next recompilation cycle
+  /// won't recompile sources that were previously reported as changed.
+  void acceptLastDelta();
+  /// Reject results of previous compilation. Next recompilation cycle will
+  /// recompile sources indicated as changed.
+  void rejectLastDelta();
+  /// This let's compiler know that source file identifed by `uri` was changed.
+  void invalidate(Uri uri);
+}
+
+class _FrontendCompiler implements CompilerInterface {
+  IncrementalKernelGenerator _ikg;
+  String _filename;
+
+  @override
+  Future<Null> compile(String filename, ArgResults options) async {
+    final String boundaryKey = new Uuid().generateV4();
+    print("result $boundaryKey");
+    final Uri sdkRoot = _ensureFolderPath(options['sdk-root']);
+    final CompilerOptions compilerOptions = new CompilerOptions()
+      ..sdkRoot = sdkRoot
+      ..strongMode = false
+      ..target = new FlutterTarget(new TargetFlags())
+      ..onError = (CompilationMessage message) {
+        print(message);
+      };
+    Program program;
+    if (options['incremental']) {
+      _ikg = await IncrementalKernelGenerator.newInstance(
+          compilerOptions, Uri.base.resolve(filename));
+      final DeltaProgram deltaProgram = await _ikg.computeDelta();
+      program = deltaProgram.newProgram;
+    } else {
+      // TODO(aam): Remove linkedDependencies once platform is directly embedded
+      // into VM snapshot and http://dartbug.com/30111 is fixed.
+      compilerOptions.linkedDependencies = <Uri>[
+        sdkRoot.resolve('platform.dill')
+      ];
+      program = await kernelForProgram(Uri.base.resolve(filename), compilerOptions);
+    }
+    if (program != null) {
+      final String kernelBinaryFilename = filename + ".dill";
+      await writeProgramToBinary(program, kernelBinaryFilename);
+      print(boundaryKey + " " + kernelBinaryFilename);
+    } else {
+      print(boundaryKey);
+    }
+    return null;
+  }
+
+  @override
+  Future<Null> recompileDelta() async {
+    _ikg.computeDelta();
+    final String boundaryKey = new Uuid().generateV4();
+    print("result $boundaryKey");
+    final DeltaProgram deltaProgram = await _ikg.computeDelta();
+    final Program program = deltaProgram.newProgram;
+    final String kernelBinaryFilename = _filename + ".dill";
+    await writeProgramToBinary(program, kernelBinaryFilename);
+    print(boundaryKey);
+    return null;
+  }
+
+  @override
+  void acceptLastDelta() {
+    _ikg.acceptLastDelta();
+  }
+
+  @override
+  void rejectLastDelta() {
+    _ikg.rejectLastDelta();
+  }
+
+  @override
+  void invalidate(Uri uri) {
+    _ikg.invalidate(uri);
+  }
+
+  Uri _ensureFolderPath(String path) {
+    if (!path.endsWith('/')) {
+      path = '$path/';
```

Comment:
Ah, excellent. Maybe leave a comment in this function saying something like "`// (This is a URL, not a path, so the forward slash is correct even on Windows.)`.